SPI broken in 2023.9.1 when 2 HW devices are used

### The problem

After upgrading from 2023.8.3 to 2023.9.1 it is not possible to compile 2 HW SPI buses.

```
src/main.cpp: In function 'void setup()':
src/main.cpp:156:28: error: expected primary-expression before 'return'
   spi_bus_2->set_interface(return new SPIClass(HSPI));
```

### Which version of ESPHome has the issue?

2023.9.1

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

NA

### What platform are you using?

ESP32

### Board

esp32dev

### Component causing the issue

SPI

### Example YAML snippet

```yaml
esphome:
  name: test
  platform: ESP32
  board: esp32dev

spi:
  - id: spi_bus_1
    clk_pin: GPIO32
    mosi_pin: GPIO33
    miso_pin: GPIO39
  - id: spi_bus_2
    clk_pin: GPIO15
    mosi_pin: GPIO13
```


### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2023.9.1
INFO Reading configuration /config/test.yaml...
WARNING GPIO15 is a Strapping PIN and should be avoided.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq.html#why-am-i-getting-a-warning-about-strapping-pins
INFO Generating C++ source...
INFO Compiling app...
Processing test (board: esp32dev; framework: arduino; platform: platformio/espressif32@5.4.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Dependency Graph
|-- SPI @ 2.0.0
Compiling .pioenvs/test/src/main.cpp.o
Compiling .pioenvs/test/FrameworkArduino/HardwareSerial.cpp.o
Compiling .pioenvs/test/FrameworkArduino/IPAddress.cpp.o
Compiling .pioenvs/test/FrameworkArduino/IPv6Address.cpp.o
Compiling .pioenvs/test/FrameworkArduino/MD5Builder.cpp.o
Compiling .pioenvs/test/FrameworkArduino/Print.cpp.o
Compiling .pioenvs/test/FrameworkArduino/Stream.cpp.o
Compiling .pioenvs/test/FrameworkArduino/StreamString.cpp.o
src/main.cpp: In function 'void setup()':
src/main.cpp:156:28: error: expected primary-expression before 'return'
   spi_bus_2->set_interface(return new SPIClass(HSPI));
                            ^~~~~~
*** [.pioenvs/test/src/main.cpp.o] Error 1
========================== [FAILED] Took 2.71 seconds ==========================
```


### Additional information

Related to https://github.com/esphome/issues/issues/4922 but I found that is not depending on devices, just the SPI bus.

So it looks like the rewrite from @clydebarrow at https://github.com/esphome/esphome/pull/5311 unfortunately broke this as it worked fine in 2023.8.3.

I also tried to switch to use the IDF framework but my device (Shelly Pro 4PM) failed. So until this is addressed devices with 2 SPI buses need to stay in 2023.8.3

--- a/esphome/components/spi/__init__.py
+++ b/esphome/components/spi/__init__.py
@@ -187,7 +187,7 @@ def get_spi_interface(index):
     # Following code can't apply to C2, H2 or 8266 since they have only one SPI
     if get_target_variant() in (VARIANT_ESP32S3, VARIANT_ESP32S2):
         return "new SPIClass(FSPI)"
-    return "return new SPIClass(HSPI)"
+    return "new SPIClass(HSPI)"
 
 
 SPI_SCHEMA = cv.All(

